source/opt/pass_manager.cpp

```diff
@@ -41,9 +49,64 @@ Pass::Status PassManager::Run(ir::IRContext* context) {
     }
   };
 
+#if defined(SPIRV_ANDROID) || defined(SPIRV_LINUX) || defined(SPIRV_MAC) || \
+    defined(SPIRV_FREEBSD)
+  double (*get_time_difference)(const timeval&, const timeval&) = NULL;
+  if (ftime_report_stream_) {
+    *ftime_report_stream_<< std::setw(30) << "Pass Name"
+      << std::setw(16) << "CPU time"
+      << std::setw(16) << "WALL time"
+      << std::setw(16) << "SYS time"
+      << std::setw(16) << "Pagefaults"
+      << std::setw(16) << "RSS" << std::endl;
+    get_time_difference = [](const timeval& before, const timeval& after) -> double {
+      return static_cast<double>(after.tv_sec - before.tv_sec) +
+        static_cast<double>(after.tv_usec - before.tv_usec) * .000001;
+    };
+  }
+#endif
   for (const auto& pass : passes_) {
     print_disassembly("; IR before pass ", pass.get());
+#if defined(SPIRV_ANDROID) || defined(SPIRV_LINUX) || defined(SPIRV_MAC) || \
+    defined(SPIRV_FREEBSD)
+    bool usage_fail = false;
+    rusage usage_before;
+    timeval wall_before;
+    if (ftime_report_stream_) {
+      if (getrusage(RUSAGE_CHILDREN, &usage_before) == -1) {
+        *ftime_report_stream_ << std::setw(30) << (pass ? pass->name() : "")
+          << " ERROR: calling getrusage() fails";
+        usage_fail = true;
+      } else if (gettimeofday(&wall_before, NULL) == -1) {
+        *ftime_report_stream_ << std::setw(30) << (pass ? pass->name() : "")
```

Comment:
Consider measuring uptime instead of calendar time.
for example clock_gettime(CLOCK_MONOTONIC)

Reply:
Thank you for letting me know it, but let me use gettimeofday() this time. I will update it if critical problems are caused.